ctapipe/tools/make_irf.py

```diff
@@ -0,0 +1,353 @@
+        SourceOffsetBinning,
+        OutputEnergyBinning,
+        EventPreProcessor,
+    ]
+
+    def setup(self):
+        self.opt_result = OptimisationResult()
+        self.theta = ThetaCutsCalculator(parent=self)
+        self.e_bins = OutputEnergyBinning(parent=self)
+        self.bins = SourceOffsetBinning(parent=self)
+        self.epp = EventPreProcessor(parent=self)
+
+        self.opt_result.read(self.cuts_file)
+        self.epp.quality_criteria = self.opt_result.precuts
+        self.reco_energy_bins = self.e_bins.reco_energy_bins()
+        self.true_energy_bins = self.e_bins.true_energy_bins()
+        self.energy_migration_bins = self.e_bins.energy_migration_bins()
+
+        self.source_offset_bins = self.bins.source_offset_bins()
+        self.fov_offset_bins = self.opt_result.offset_lim
+
+        self.particles = [
+            EventSelector(
+                self.epp,
+                "gammas",
+                self.gamma_file,
+                PYIRF_SPECTRA[self.gamma_sim_spectrum],
+            ),
+            EventSelector(
+                self.epp,
+                "protons",
+                self.proton_file,
+                PYIRF_SPECTRA[self.proton_sim_spectrum],
+            ),
+            EventSelector(
+                self.epp,
+                "electrons",
+                self.electron_file,
+                PYIRF_SPECTRA[self.electron_sim_spectrum],
+            ),
+        ]
+
+    def start(self):
+        reduced_events = dict()
+        for sel in self.particles:
+            evs, cnt = sel.load_preselected_events(self.chunk_size)
+            reduced_events[sel.kind] = evs
+            reduced_events[f"{sel.kind}_count"] = cnt
+
+        self.log.debug(
+            "Loaded %d gammas, %d protons, %d electrons"
+            % (
+                reduced_events["gamma_count"],
+                reduced_events["proton_count"],
+                reduced_events["electron_count"],
+            )
+        )
+
+        self.signal_events = reduced_events["gamma"]
+        self.background_events = vstack(
+            [reduced_events["proton"], reduced_events["electron"]]
+        )
+        self.signal_events["selected_gh"] = evaluate_binned_cut(
+            self.signal_events["gh_score"],
+            self.signal_events["reco_energy"],
+            self.opt_result.gh_cuts,
+            operator.ge,
+        )
+        self.background_events["selected_gh"] = evaluate_binned_cut(
+            self.background_events["gh_score"],
+            self.background_events["reco_energy"],
+            self.opt_result.gh_cuts,
+            operator.ge,
+        )
+        self.theta_cuts_opt = self.theta.calculate_theta_cuts(
+            self.signal_events[self.signal_events["selected_gh"]]["theta"],
+            self.signal_events[self.signal_events["selected_gh"]]["reco_energy"],
+            self.reco_energy_bins(),
+        )
+
+        self.signal_events["selected_theta"] = evaluate_binned_cut(
```

Comment:
Applying the theta cut should be optional, we don't need it for computing general (not point-like) IRFs.